kv/kvserver: TestRaftCheckQuorum failed

kv/kvserver.TestRaftCheckQuorum [failed](https://tanzanite.cluster.engflow.com/invocations/default/0cf35bfc-d5ae-4b97-8608-920c2c5894d4?testReportRun=15&testReportShard=48&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [f717f6bd218121bb5e3376af658545f6bff30c22](https://github.com/cockroachdb/cockroach/commits/f717f6bd218121bb5e3376af658545f6bff30c22):


```
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 2: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 3: to 1
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StatePreCandidate at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_helpers_test.go:475: allowing msg MsgHeartbeat from store 1: to 1
    client_raft_helpers_test.go:475: allowing msg MsgHeartbeatResp from store 1: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeatResp from store 2: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeatResp from store 3: to 1
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StatePreCandidate at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 2: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 3: to 1
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_helpers_test.go:475: allowing msg MsgHeartbeat from store 1: to 1
    client_raft_helpers_test.go:475: allowing msg MsgHeartbeatResp from store 1: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeatResp from store 3: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeatResp from store 2: to 1
    client_raft_test.go:6565: n2 StatePreCandidate at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n2 StateFollower at term=6 commit=25
    client_raft_test.go:6565: n3 StateFollower at term=6 commit=25
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 2: to 1
    client_raft_helpers_test.go:473: dropping msg MsgHeartbeat from store 3: to 1
    client_raft_test.go:6563: 
        	Error Trace:	pkg/kv/kvserver/client_raft_test.go:6563
        	            				pkg/testutils/subtest.go:23
        	Error:      	Condition never satisfied
        	Test:       	TestRaftCheckQuorum/symmetric=false
    --- FAIL: TestRaftCheckQuorum/symmetric=false (20.39s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=15</code>
 - <code>shard=48</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftCheckQuorum.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/raft/raft.go
+++ b/pkg/raft/raft.go
@@ -2127,9 +2127,11 @@ func stepLeader(r *raft, m pb.Message) error {
 // stepCandidate is shared by StateCandidate and StatePreCandidate; the difference is
 // whether they respond to MsgVoteResp or MsgPreVoteResp.
 func stepCandidate(r *raft, m pb.Message) error {
-	if IsMsgFromLeader(m.Type) {
+	if IsMsgFromLeader(m.Type) && m.Type != pb.MsgDeFortifyLeader {
 		// If this is a message from a leader of r.Term, transition to a follower
 		// with the sender of the message as the leader, then process the message.
+		// One exception is MsgDeFortifyLeader where it doesn't mean that there is
+		// currently an active leader for this term.
 		assertTrue(m.Term == r.Term, "message term should equal current term")
 		r.becomeFollower(m.Term, m.From)
 		return r.step(r, m) // stepFollower

--- a/pkg/raft/raft_test.go
+++ b/pkg/raft/raft_test.go
@@ -2488,6 +2488,60 @@ func testDisruptiveFollowerPreVote(t *testing.T, storeLivenessEnabled bool) {
 	require.Equal(t, pb.StateLeader, n1.state)
 }
 
+// TestPreCandidateIgnoresDefortification tests that a pre-candidate ignores
+// MsgDefortifyLeader and doesn't become a follower again.
+func TestPreCandidateIgnoresDefortification(t *testing.T) {
+	var fabric *raftstoreliveness.LivenessFabric
+	var n1, n2 *raft
+
+	fabric = raftstoreliveness.NewLivenessFabricWithPeers(1, 2)
+	n1 = newTestRaft(1, 10, 1, newTestMemoryStorage(withPeers(1, 2)),
+		withStoreLiveness(fabric.GetStoreLiveness(1)))
+	n2 = newTestRaft(2, 10, 1, newTestMemoryStorage(withPeers(1, 2)),
+		withStoreLiveness(fabric.GetStoreLiveness(2)))
+
+	n1.checkQuorum = true
+	n2.checkQuorum = true
+	n1.preVote = true
+	n2.preVote = true
+
+	n1.becomeFollower(1, None)
+	n2.becomeFollower(1, None)
+
+	nt := newNetworkWithConfigAndLivenessFabric(nil, fabric, n1, n2)
+
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	// Check raft states.
+	require.Equal(t, pb.StateLeader, n1.state)
+	require.Equal(t, pb.StateFollower, n2.state)
+
+	// The term is 2 for both nodes.
+	require.Equal(t, uint64(2), n1.Term)
+	require.Equal(t, uint64(2), n2.Term)
+
+	// Withdraw 2's support for 1. This allows 2 to pre-campaign since it's not
+	// supporting a fortified leader.
+	nt.livenessFabric.WithdrawSupportFor(2, 1)
+
+	// Isolate 1 so that it doesn't receive the MsgVoteRequest from 2, and
+	// therefore it doesn't vote for it.
+	nt.isolate(1)
+	nt.send(pb.Message{From: 2, To: 2, Type: pb.MsgHup})
+
+	// 2 is now a pre-candidate.
+	require.Equal(t, pb.StatePreCandidate, n2.state)
+
+	// 2 should remain a PreCandidate even if it receives a MsgDefortifyLeader.
+	nt.send(pb.Message{From: 1, To: 2, Term: 2, Type: pb.MsgDeFortifyLeader})
+	require.Equal(t, pb.StatePreCandidate, n2.state)
+
+	// However, receiving another message from a leader would cause 2 to become
+	// follower again.
+	nt.send(pb.Message{From: 1, To: 2, Term: 2, Type: pb.MsgApp})
+	require.Equal(t, pb.StateFollower, n2.state)
+}
+
 func TestLeaderAppResp(t *testing.T) {
 	// initial progress: match = 0; next = 3
 	tests := []struct {

